Metadata API: include target target name in TargetFile

MVrachev opened the pull request:
Related to #1411, but it's not yet clear if it closes it. 

**Description of the changes being introduced by the pull request**:

Currently, TargetFile instances do not contain the filename of the file
they represent. The API itself does not need it but it could be useful
for users of the API.

As an example, the current client returns a dict for
`get_one_valid_targetinfo()`: that dict contains a filepath field and
a targetinfo field (essentially TargetFile).
We would like to keep a similar API, but avoid hand-crafted dicts.
It would be much nicer to return a TargetFile that would contain the
full "metadata" of the targetfile.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Metadata API: include target path in targetfile

Currently, TargetFile instances do not contain the path relative URL of
the file they represent. The API itself does not need it but it could be
useful for users of the API.

As an example, the current client returns a dict for
get_one_valid_targetinfo(): that dict contains a filepath field and
a targetinfo field (essentially TargetFile).
We would like to keep a similar API, but avoid hand-crafted dicts.
It would be much nicer to return a TargetFile that would contain the
full "metadata" of the targetfile.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

sechkova commented on tuf/ngclient/updater.py:
```diff
@@ -460,7 +459,8 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
             )
 
-        return {"filepath": target_filepath, "fileinfo": target}
+        target.targetname = target_filepath
```
This doesn't feel right. This is not the place to set "targetname". It should be done during the instantiation of a Targets object.

MVrachev replied:
I had to do that before when `targetname` was an optional attribute.
Now, when it's mandatory there is no need to do such a thing.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -1085,13 +1085,15 @@ class TargetFile(BaseFile):
         self,
         length: int,
         hashes: Dict[str, str],
+        targetname: Optional[str] = None,
```
I'm not a fan of`TargetFile.targetname`. How about `TargetFile.filename`, `TargetFile.path` or `TargetFile.targetpath` if we insist on being verbose and stick to the spec.

jku replied:
(this is just a drive by comment, will have to think about the justification for thw whole thing a bit still... but thanks for starting this, I think it's useful to make discussion more concrete)

What teodora said,  let's not use even more different names for the same thing... I think "targetpath" is used for this already so the attribute should probably be targetpath or path.

I believe it should not be optional though. Or is there a use case of a TargetFile without a targetpath?

MVrachev replied:
> I'm not a fan of TargetFile.targetname. How about TargetFile.filename, TargetFile.path or TargetFile.targetpath if we insist on being verbose and stick to the spec.

I had a misconception and didn't want to call it `targetpath` thinking it's not an actual path, but a file name.
Now, I read the spec and it clearly this is called `TARGETPATH`:
https://theupdateframework.github.io/specification/latest/#targetpath.

So, from both options to call it `targetpath` or `path` I would prefer `path` given it's shorter.

> I believe it should not be optional though. Or is there a use case of a TargetFile without a targetpath?

I wasn't sure when I added it as optional. I did that because it's not required by the spec, but we can use `TargetFile`
in such a way that we would always pass the path.

So, in conclusion, I will rename the attribute to `path` and make it non-optional.

jku commented:
I think this makes sense and implementation looks fine. You will have to do something about `Targets.update()` though (I still think removing it is correct -- at least currently it seems to have no purpose other than creating apparent complexity -- but I guess fixing it to match the new API is an option as well)

Maybe it would be good to document target path a little more: that it is a "path-relative-URL string" that will be used as part of the final download url to fetch the target

The high level question is whether anyone objects to making objects like this a bit smarter (in this case be aware of the path) even if the file format does not include that. I think this is a good way forward and seems immediately understandable by the reader -- it definitely makes the Updater API much better...

jku commented:
For those not following that closely: this PR changes TargetFile interface (which on it's own is a minor change in the Metadata API implementation) but also the whole ngclient API so that TargetFile is returned and expected as argument instead of an undefined dictionary.

I think this is going in the right direction and allows us to continue with something like https://github.com/theupdateframework/tuf/issues/1317#issuecomment-843872644: I believe the core Updater should not handle caching: this would make it much clearer what is actually part of the spec implementation and what is an additional thing we just wanted to implement in the same source directory

joshuagl commented:
> The high level question is whether anyone objects to making objects like this a bit smarter (in this case be aware of the path) even if the file format does not include that. I think this is a good way forward and seems immediately understandable by the reader -- it definitely makes the Updater API much better...

I do not object, in fact looking at it now having an object which describes a file without being able to explicitly identify the file does look odd. +1 from me.

jku commented:
@MVrachev can you suggest a solution for `Targets.update()`: at the very least the function arguments need to be updated.

jku reviewed (changes requested):
marking changes requested for `Targets.update()` issue

Commit:
Remove filename argument from Targets.update()

After the addition of "path" argument in the TargetFile class the
filename argument in Targets.update() became redundant.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I had to rebase to fix merge conflicts after the simplification of `_preorder_depth_first_walk`.
I updated the `Targets.update()` argument accordingly.
I will prefer if other changes or removal of the update method are done for all classes when resolving #1230.

jku commented on tuf/ngclient/updater.py:
```diff
@@ -252,27 +249,26 @@ def download_target(
 
-        target_fileinfo: "TargetFile" = targetinfo["fileinfo"]
-        target_filepath = targetinfo["filepath"]
+        target_filepath = targetinfo.path
```
nitpick but this variable seems unneeded

MVrachev replied:
If we remove target_filepath we would have to define `full_url` before the `if` statement so it can be used
with and without `consistent_snapshots` and `prefix_targets_with_hash`.
It will look something like that
```python
        full_url = parse.urljoin(target_base_url, targetinfo.path)
        consistent_snapshot = self._trusted_set.root.signed.consistent_snapshot
        if consistent_snapshot and self.config.prefix_targets_with_hash:
            hashes = list(targetinfo.hashes.values())
            full_url = parse.urljoin(
                target_base_url, f"{hashes[0]}.{targetinfo.path}"
            )
```

Also, when raising the exception on line 265 we use the `target_filepath` which could be with HASH prefix or without and the correct target name will be saved in `target_filepath`.

jku replied:
thanks, makes sense

jku commented on tuf/ngclient/updater.py:
```diff
@@ -414,7 +410,8 @@ def _preorder_depth_first_walk(
             if target is not None:
                 logger.debug("Found target in current role %s", role_name)
-                return {"filepath": target_filepath, "fileinfo": target}
+                target.targetname = target_filepath
```
I don't understand this. Why is the updater modifying a TargetFile?

MVrachev replied:
You are right, that's useless.
Totally forgot that the TargetFile instance itself will have the correct target path already.
PS: I noticed I haven't changed `targetname` to `path` after we decide to rename it and no test had failed.
We really need tests for that functionality...

jku replied:
it can't be a real test (as defining new attributes like this is legal in python) but _maybe_ mypy has an option to warn about defining object attributes outside init() :shrug:

jku commented on tuf/api/metadata.py:
```diff
@@ -1138,13 +1138,15 @@ class TargetFile(BaseFile):
     Attributes:
         length: An integer indicating the length of the target file.
         hashes: A dictionary of hash algorithm names to hash values.
+        path: A string denoting the target file path.
```
I'd like to not call it a _file path_ since at least for me it creates the impression that we are talking about local filesystem paths: target path is really a url fragment and is used to form the target download URL. Our updater does end up caching it locally using the target path as a file path, but that's an implementation detail IMO.

Maybe we could mention "path-relative-URL string" like https://theupdateframework.github.io/specification/latest/#targetpath does? not sure it's any more descriptive but it would avoid creating the impression of local path?

jku reviewed (commented):
Looks nice. I left a couple of comments and a question.

jku commented on tuf/api/metadata.py:
```diff
@@ -1138,13 +1138,16 @@ class TargetFile(BaseFile):
         length: An integer indicating the length of the target file.
         hashes: A dictionary of hash algorithm names to hash values.
+        path: A string denoting the path to a target file relative to a base
+            URL of targes.
```
typo of targets
